modules/rtm/R/bayestools.R

```diff
@@ -0,0 +1,168 @@
+#' Generic log-likelihood generator for RTMs
+rtm_loglike <- function(nparams, model, observed, lag.max = 0.01, ...) {
+    fail_ll <- -1e10
+    stopifnot(nparams >= 1, nparams %% 1 == 0, is.function(model), is.numeric(observed))
+    n_obs <- length(observed)
+    out <- function(x) {
+        rtm_params <- x[seq_len(nparams)]
+        rsd <- x[nparams + 1]
+        mod <- model(rtm_params, ...)
+        if (any(is.na(mod))) return(fail_ll)
+        err <- mod - observed
+        ss <- sum(err * err)
+        sigma2 <- rsd * rsd
+        n_eff <- neff(err, lag.max = lag.max)
+        sigma2eff <- sigma2 * n_obs / n_eff
+        ll <- -0.5 * (n_obs * log(sigma2eff) + ss / sigma2eff)
+        if (is.na(ll)) return(fail_ll)
+        return(ll)
+    }
+    return(out)
+}
+
+#' Check convergence of BayesianTools output
+bt_check_convergence <- function(samples, threshold = 1.1, use_CI = TRUE, use_mpsrf = TRUE) {
+    i <- ifelse(use_CI, 2, 1)
+    gelman <- BayesianTools::gelmanDiagnostics(samples)
+    if (use_mpsrf) {
+        gelman_vec <- c(gelman$psrf[,i], mpsrf = gelman$mpsrf)
+    } else {
+        gelman_vec <- gelman$psrf[,i]
+    }
+    exceeds <- gelman_vec > threshold
+    if (any(exceeds)) {
+        exceeds_vec <- gelman_vec[exceeds]
+        exceeds_char <- sprintf('%s: %.2f', names(exceeds_vec), exceeds_vec)
+        exceeds_str <- paste(exceeds_char, collapse = '; ')
+        message('The following parameters exceed threshold: ', exceeds_str)
+        return(FALSE)
+    } else {
+        return(TRUE)
+    }
+}
+
+#' Quick BayesianTools prior creator for PROSPECT model 
+#' 
+#' @param custom_prior List containing `param_name`, `distn`, `parama`, `paramb`, and `lower`
+#' @inheritParams prospect
+#' @export
+prospect_bt_prior <- function(version, custom_prior = list()) {
+    col_names <- c('param_name', 'distn', 'parama', 'paramb', 'lower')
+    prior_default_list <- list(N = list('N', 'norm', 1.4, 0.8, 1),
+                               Cab = list('Cab', 'lnorm', log(40), 0.9, 0),
+                               Car = list('Car', 'lnorm', log(10), 1.1, 0),
+                               Cbrown = list('Cbrown', 'lnorm', log(1), 1.1, 0),
+                               Cw = list('Cw', 'lnorm', log(0.01), 1, 0),
+                               Cm = list('Cm', 'lnorm', log(0.009), 1, 0),
+                               residual = list('residual', 'lnorm', log(0.001), 2.5, 0)
+                               )    
+    prior_list <- modifyList(prior_default_list, custom_prior)
+    prior_df_all <- do.call(rbind.data.frame, prior_list)
+    colnames(prior_df_all) <- col_names
+    default_params <- defparam(paste0('prospect_', tolower(version)))
```

Comment:
@ashiklom is defparam a default P RTM function to return a vector of default parameter values?